### DISCUSSION TITLE
Are recursive sum types supported?

### DISCUSSION THREAD

**Discussion by avillega** [vlang/v#7932] [General]
Hi, I am trying to implement a tiny lisp using V I kind of came up with this definition for an Expression sum type 
```
enum Op {
	plus
	minus
	times
	divide
}

type Expr = Op | f64 | []Expr
```

When using this I will need match this way 
```
match e {
  f64 {
  ...
  }
  Op {
  ...
  }
  []Expr {
  ...
  }
}
```
However the compiler is getting confused thinking that the []Expr branch in the match statement is initializing an array or a map.

What would be an idiomatic way of implementing something like this in V.

thank you!

**Comment by medvednikov** [comment]
We've never thought of using it like this, but I guess there's no reason not to allow this.

This should be possible in the new sum type implementation, right @danieldaeschle?

**Reply by avillega** [reply]
BTW the compiler does not complaint in the definition, in fact if I use a `else` branch instead of the `[]Expr` branch it works with the recursive definition. If I don't add a `[]Expr` branch the compiler tells me to add that branch or an `else` branch 
```
lisv.v:29:2: error: match must be exhaustive (add match branches for: `[]Expr` or `else {}` at the end)
   27 | fn eval(expr []Expr) Expr {
   28 |     e := expr[0]
   29 |     match e {
      |     ~~~~~~~~~
   30 |         f64 {
   31 |             return e
```

**Reply by danieldaeschle** [reply]
yep, should be possible i think. at least from the language part. I will see what i can do.

**Reply by danieldaeschle** [reply]
@avillega could you create this as an issue with a minimal example?

**Reply by danieldaeschle** [reply]
https://github.com/vlang/v/pull/7939

**Reply by avillega** [reply]
Sorry I couldn't create the issue yesterday it was late here in Canada

**Reply by avillega** [reply]
Just in case it is still useful here is the issue #7948

**Reply by medvednikov** [reply]
@avillega it's already been fixed by @danieldaeschle :)

**Reply by danieldaeschle** [reply]
Could you add a category called "Question" to discussions?
